libass/ass_drawing.c

```diff
@@ -43,93 +43,149 @@ static bool token_check_values(ASS_DrawingToken *token, int i, ASS_TokenType typ
-            if (root) {
-                tail->next = calloc(1, sizeof(ASS_DrawingToken));
-                tail->next->prev = tail;
-                tail = tail->next;
-            } else {
-                /* VSFilter compat:
-                 * In guliverkli(2) VSFilter all drawings
-                 * whose first valid command isn't m are rejected.
-                 * xy-VSF and MPC-HC ISR this was (possibly inadvertenly) later relaxed,
-                 * such that all valid commands but n are ignored if there was no m yet.
-                 */
-                if (type == TOKEN_MOVE_NC) {
-                    return NULL;
-                } else if (type != TOKEN_MOVE) {
-                    p++;
+        char cmd = *p;
+        p++;
+        /* VSFilter compat:
+         * In guliverkli(2) VSFilter all drawings
+         * whose first valid command isn't m are rejected.
+         * xy-VSF and MPC-HC ISR this was (possibly inadvertenly) later relaxed,
+         * such that all valid commands but n are ignored if there was no m yet.
+         * 'p' is further restricted and ignored unless there are already >= 3 nodes
+         * in newer VSFs, but guliverkli2 just crashed on 'p' with  1 <= n < 3 preceding nodes.
+         */
+        switch (cmd) {
+        case 'm':
+            if (!root) {
+                ASS_Vector coord;
+                if (!get_coordinate(&p, &coord))
                     continue;
-                }
                 root = tail = calloc(1, sizeof(ASS_DrawingToken));
+                root->point = coord;
+                root->type = TOKEN_MOVE;
+                coords = 1;
+            }
+            coords += add_coordinates(&p, &tail, TOKEN_MOVE, 1, 1);
+            break;
+        case 'n':
+            if (!root)
+                return NULL;
+            coords += add_coordinates(&p, &tail, TOKEN_MOVE_NC, 1, 1);
+            break;
+        case 'l':
+            if (!root)
+                continue;
+            coords += add_coordinates(&p, &tail, TOKEN_LINE, 1, 1);
+            break;
+        case 'b':
+            if (!root)
+                continue;
+            coords += add_coordinates(&p, &tail, TOKEN_CUBIC_BEZIER, 1, 3);
+            break;
+        case 's':
+            if (!root)
+                continue;
+            spline_start = tail;
+            size_t new_coords = add_coordinates(&p, &tail, TOKEN_B_SPLINE, 3, 1);
+            coords += new_coords;
+            if (new_coords == 0)
+                spline_start = NULL;
+            break;
+        case 'p':
+            if (coords < 3)
+                continue;
+            // XXX: use TOKEN_EXTEND_SPLINE
+            coords += add_coordinates(&p, &tail, TOKEN_B_SPLINE, 1, 1);
+            break;
+        case 'c':
+            if (!spline_start)
+                continue;
+            // Close b-splines: add the first three points of the b-spline
+            // back to the end (VSF does not check the original node types)
```

Comment:
If I’m reading the code (here and in VSFilter) correctly, then I’m not sure what this parenthetical comment is supposed to mean: the second and third point are guaranteed to be `TOKEN_B_SPLINE`, while the first point can legally be anything, because it’s the point _before_ the spline rather than part of the spline command. But there can be non-spline nodes between `spline_start` and the `c` command, which may be surprising. Well, the whole `c` command seems weird to me, but what can you do.

Reply:
You are right, this part of the comment is kinda useless; dropped it.  
And yeah, `c` can ""close"" non-splines (there might even have been an `m` inbetween) and `p` also can ""extend"" non-splines